internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -0,0 +1,270 @@
+		latencyMultiplier:          latencyMultiplier,
+		maxExpectedWantProcessTime: maxExpectedWantProcessTime,
+		onDontHaveTimeout:          onDontHaveTimeout,
+	}
+
+	return mqp
+}
+
+// Shutdown the dontHaveTimeoutMgr. Any subsequent call to Start() will be ignored
+func (dhtm *dontHaveTimeoutMgr) Shutdown() {
+	dhtm.shutdown()
+}
+
+// onShutdown is called when the dontHaveTimeoutMgr shuts down
+func (dhtm *dontHaveTimeoutMgr) onShutdown() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Clear any pending check for timeouts
+	if dhtm.checkForTimeoutsTimer != nil {
+		dhtm.checkForTimeoutsTimer.Stop()
+	}
+}
+
+// onStarted is called when the dontHaveTimeoutMgr starts.
+// It monitors for the context being cancelled.
+func (dhtm *dontHaveTimeoutMgr) onStarted() {
+	<-dhtm.ctx.Done()
+	dhtm.onShutdown()
+}
+
+// Start the dontHaveTimeoutMgr. This method is idempotent
+func (dhtm *dontHaveTimeoutMgr) Start() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Make sure the dont have timeout manager hasn't already been started
+	if dhtm.started {
+		return
+	}
+	dhtm.started = true
+
+	go dhtm.onStarted()
+
+	// If we already have a measure of latency to the peer, use it to
+	// calculate a reasonable timeout
+	latency := dhtm.peerConn.Latency()
+	if latency.Nanoseconds() > 0 {
+		dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+		return
+	}
+
+	// Otherwise measure latency by pinging the peer
+	go dhtm.measureLatency()
+}
+
+// measureLatency measures the latency to the peer by pinging it
+func (dhtm *dontHaveTimeoutMgr) measureLatency() {
+	// Wait up to defaultTimeout for a response to the ping
+	ctx, cancel := context.WithTimeout(dhtm.ctx, dhtm.defaultTimeout)
+	defer cancel()
+
+	// Ping the peer
+	res := dhtm.peerConn.Ping(ctx)
+	if res.Error != nil {
+		// If there was an error, we'll just leave the timeout as
+		// defaultTimeout
+		return
+	}
+
+	// Get the average latency to the peer
+	latency := dhtm.peerConn.Latency()
+
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Calculate a reasonable timeout based on latency
+	dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+
+	// Check if after changing the timeout there are any pending wants that are
+	// now over the timeout
+	go dhtm.checkForTimeouts()
+}
+
+// checkForTimeouts checks pending wants to see if any are over the timeout
+func (dhtm *dontHaveTimeoutMgr) checkForTimeouts() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	if len(dhtm.pending) == 0 {
+		return
+	}
+
+	// Figure out which of the blocks that were wanted were not received
+	// within the timeout
+	pending := make([]cid.Cid, 0, len(dhtm.pending))
+	var oldest time.Time
+	for c, started := range dhtm.pending {
+		if time.Since(started) > dhtm.timeout {
+			pending = append(pending, c)
+			delete(dhtm.pending, c)
+		} else if oldest.IsZero() || started.Before(oldest) {
+			oldest = started
+		}
+	}
+
+	// Fire the timeout event for the pending wants
+	if len(pending) > 0 {
+		go dhtm.fireTimeout(pending)
```

Comment:
Same, does this need a goroutine?

Reply:
For this one we may want to keep it in a go-routine:
The event handler will call the WantManager with a DONT_HAVE, which causes a whole bunch of other stuff to happen, and if it's not in a go routine it may end up blocking the timeout manager for a while, and possibly causing a race condition.

Reply:
> and possibly causing a race condition

Race condition or a deadlock? If blocking causes a race condition, there's a bug already.

Otherwise, SGTM. I'm actually a fan of slowing things down in the name of backpressure. That is, if want manager is slow, we should stop sending it stuff. However, if that could cause a deadlock, that's an entirely different story.

Reply:
Sorry I meant deadlock. I agree with the back pressure argument in general. In this case the back pressure would be between the client side of the node and its own server side (rather than a remote peer's server side), which feels a little weird and deadlocky. In future we may decide to handle this onTimeout event in a different way, so to be more future-proof my vote is to keep the go routine.

Reply:
Ah, makes sense.

Reply:
(I just want to make sure we don't buffer with goroutines, we've done that before and it's _expensive_).